### PR TITLE
Introduce the `ContainerdRegistryHostsDir` feature gate

### PR DESCRIPTION
**Pull request by ialidzhikov**
**How to categorize this PR?**

/area cost
/kind enhancement

**What this PR does / why we need it**:
Currently there are two configuration patterns for configuring containerd registries. As part of https://github.com/gardener/gardener-extension-registry-cache/issues/3, for the registry-cache extension we decided to use the new and recommended configuration pattern - [hosts directory pattern](https://github.com/containerd/containerd/blob/main/docs/hosts.md).
Only one of the configuration patterns has to be followed strictly. In order to realise the hosts directory pattern in the registry-cache, we have to adapt registry config contributed by provider-local to be in hosts directory pattern.
We decided gardenlet (containerd-initializer unit's script) to take over to create the `/etc/containerd/certs.d` dir and to add the following config to containerd's config.toml:
```toml
[plugins."io.containerd.grpc.v1.cri".registry]
   config_path = "/etc/containerd/certs.d"
```

Providing such config should be okay for Shoots that don't want to configure containerd registries and should not have any side effects.
The motivation to introduce this handling in gardenlet (containerd-initializer unit's script):
- Expose the containerd registry configuration as feature (similar to the containerd's import feature). End users need to created only dir and hosts.toml file under `/etc/containerd/certs.d` to configure containerd registries.
- Make it easier for extensions (provider-local and gardener-extension-registry-cache) to contribute registry config. With this feature, such extensions only have to mutate the OperatingSystemConfig resource and add the corresponding config files there. And they don't need to add systemd units that append/remove containerd config.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener-extension-registry-cache/issues/3

**Special notes for your reviewer**:
N/A

**Release note**:

```feature user
A new feature gate named `ContainerdRegistryHostsDir` is introduced to gardenlet. When enabled, the `/etc/containerd/certs.d` directory is created on the Node and containerd is configured to look up for registries/mirrors configuration in this directory (if there is any configuration applied). In future, the [registry-cache extension](https://github.com/gardener/gardener-extension-registry-cache/) will add such registries/mirrors configuration under this directory (via OperatingSystemConfig mutation).
```

### CODE REVIEW

File: pkg/provider-local/webhook/controlplane/ensurer.go [hunk 1]
```diff
@@ -144,6 +145,31 @@ func (e *ensurer) EnsureAdditionalFiles(ctx context.Context, gc extensionscontex
...
+		appendFileIfNotPresent(new, extensionsv1alpha1.File{
+			Path:        filepath.Join("/etc/containerd/certs.d", mirror.UpstreamHost, "hosts.toml"),
+			Permissions: pointer.Int32(0644),
+			Content: extensionsv1alpha1.FileContent{
+				Inline: &extensionsv1alpha1.FileContentInline{
+					Encoding: "",
+					Data:     mirror.HostsTOML(),
+				},
+			},
+		})
```
Comment (ialidzhikov) [inline pkg/provider-local/webhook/controlplane/ensurer.go]: @rfranzke , I tried to raise this topic earlier today. The motivation for this PR is not only a step in enabling registry-cache (the config_path field gets managed by gardenlet) but this PR also adapts provider-local to use the new config pattern. Hence, it will be possible to enable registry-cache extension using the new config pattern for a Shoot in local setup. 
The problem is that provider-local and registry-cache (if enabled and configured in local setup) will try to append the same files but with different config:
- provider-local will configure the registries running in the kind cluster
- registry-cache will configure the registries that it deployed to the Shoot

For now, I adapted provider-local to append file to the OSC only if it is missing. In this way, if file/config is already appended to the OSC by the registry-cache extension, it will take precedence. This will allow testing in local setup for the registry-cache extension. The drawback for the provider-local is that if someone does a code change to the appended files, they won't be applied for existing Shoot. It is not a big concern but I had doubts in general for this approach.

I was also thinking for an annotation for a Shoot - when set, provider-local to do not configure registries. This will allow registry-cache devs to use it.

cc @timebertt @oliver-goetz

Comment (ialidzhikov) [reply pkg/provider-local/webhook/controlplane/ensurer.go]: For now I left it as it is and added an explanation as comment. If there are better suggestions, I would be happy to adapt.

**Comment by acumino** [comment]
For some reason, tide is not executed on this PR. /cc @oliver-goetz

**Comment by ialidzhikov** [comment]
/hold
I noticed one issue

**Comment by ialidzhikov** [comment]
/unhold
the above-mentioned issue is now resolved

File: pkg/component/extensions/operatingsystemconfig/original/components/containerd/templates/scripts/init.tpl.sh [hunk 2]
```diff
@@ -11,6 +11,37 @@ sandbox_image_line="$(grep sandbox_image $FILE | sed -e 's/^[ ]*//')"
...
+  cat <<EOF >> $FILE
+[plugins."io.containerd.grpc.v1.cri".registry] # gardener-managed
+  config_path = "/etc/containerd/certs.d"
+EOF
+else
+  if grep --quiet --fixed-strings '[plugins."io.containerd.grpc.v1.cri".registry] # gardener-managed' "$FILE"; then
+    echo "CRI registry section is already gardener managed. Nothing to do."
+  else
+    echo "CRI registry section is not gardener managed. Setting config_path = \"$CONFIG_PATH\" in $FILE."
+    sed --null-data --in-place 's/\(\[plugins\."io\.containerd\.grpc\.v1\.cri"\.registry\]\)\n\(\s*config_path = \)""\n/\1 \# gardener-managed\n\2"\/etc\/containerd\/certs.d"\n/' "$FILE"
```
Comment (oliver-goetz) [inline pkg/component/extensions/operatingsystemconfig/original/components/containerd/templates/scripts/init.tpl.sh]:
```suggestion
    sed --null-data --in-place 's/\(\[plugins\."io\.containerd\.grpc\.v1\.cri"\.registry\]\)\n\(\s*config_path\s*=\s*\)""\n/\1 \# gardener-managed\n\2"\/etc\/containerd\/certs.d"\n/' "$FILE"
```

File: docs/usage/containerd-registry-configuration.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+> Note: The below-described functionality is provided by the `ContainerdRegistryHostsDir` feature gate in gardenlet.
+
+Gardener supports configuring `containerd` registries on a Shoot using the new [hosts directory pattern](https://github.com/containerd/containerd/blob/main/docs/hosts.md). For each Shoot Node, Gardener creates the `/etc/containerd/certs.d` directory and adds the following section to the containerd's `/etc/containerd/config.toml` file:
+```toml
+[plugins."io.containerd.grpc.v1.cri".registry] # gardener-managed
+   config_path = "/etc/containerd/certs.d"
+```
+
+In this way, end users can use the [hosts directory pattern](https://github.com/containerd/containerd/blob/main/docs/hosts.md) to configure registries for containerd.
```
Comment (oliver-goetz) [inline docs/usage/containerd-registry-configuration.md]:
```suggestion
In this way, end users can use the [hosts directory pattern](https://github.com/containerd/containerd/blob/main/docs/hosts.md) to configure registries for containerd. For doing this, shoot owners should create those files on the nodes.
```

**Review by oliver-goetz — COMMENTED** [review]
@dimitar-kostadinov @Kostov6 and me had a joined review on this PR.

Very good PR, thanks 🚀 
we found some nits only.

File: docs/README.md [hunk 4]
```diff
@@ -27,6 +27,7 @@
 * [Audit a Kubernetes cluster](usage/shoot_auditpolicy.md)
 * [Auto-Scaling for shoot clusters](usage/shoot_autoscaling.md)
 * [Cleanup of Shoot clusters in deletion](usage/shoot_cleanup.md)
+* [`containerd` registry configuration](usage/containerd-registry-configuration.md)
```
Comment (rfranzke) [inline docs/README.md]:
```suggestion
* [`containerd` Registry Configuration](usage/containerd-registry-configuration.md)
```

File: docs/usage/containerd-registry-configuration.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
+---
+title: containerd Registry Configuration
+---
+
+# `containerd` Registry Configuration
+
+containerd supports configuring registries and mirrors. Using this native containerd feature Shoot owners can configure containerd to use public or private mirrors for a given upstream registry. More details about the registry configuration can be found in the [corresponding upstream documentation](https://github.com/containerd/containerd/blob/main/docs/hosts.md).
```
Comment (rfranzke) [inline docs/usage/containerd-registry-configuration.md]:
```suggestion
containerd supports configuring registries and mirrors. Using this native containerd feature, Shoot owners can configure containerd to use public or private mirrors for a given upstream registry. More details about the registry configuration can be found in the [corresponding upstream documentation](https://github.com/containerd/containerd/blob/main/docs/hosts.md).
```

File: docs/usage/containerd-registry-configuration.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+### `containerd` Registry Configuration Patterns
+
+At the time of writing this document, containerd support two patterns for configuring registries/mirrors.
+
+> Note: Trying to use both of the patterns at the same time is not supported by containerd. Only one of the configuration patterns has to be followed strictly.
+
+##### Old and Deprecated Pattern
+
+The old and deprecated pattern is for specifying `registry.mirrors` and `registry.configs` in the containerd's config.toml file. See the [upstream documentation](https://github.com/containerd/containerd/blob/main/docs/cri/registry.md).
```
Comment (rfranzke) [inline docs/usage/containerd-registry-configuration.md]:
```suggestion
The old and deprecated pattern is specifying `registry.mirrors` and `registry.configs` in the containerd's config.toml file. See the [upstream documentation](https://github.com/containerd/containerd/blob/main/docs/cri/registry.md).
```

File: docs/usage/containerd-registry-configuration.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+Example of the old and deprecated pattern:
+```toml
```
Comment (rfranzke) [inline docs/usage/containerd-registry-configuration.md:20]:
```suggestion

```toml
```

File: docs/usage/containerd-registry-configuration.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+```
+
+In the above example, containerd is configured to first try to pull `docker.io` images from a configured endpoint (`https://public-mirror.example.com`). If the image is not available in `https://public-mirror.example.com`, then containerd will fall back to the upstream registry (`docker.io`) and will pull the image from there.
+
+##### Hosts Directory Pattern
+
+The hosts directory pattern is the new and recommended pattern for configuring registries. It is available starting `containerd@v1.5.0`. See the [upstream documentation](https://github.com/containerd/containerd/blob/main/docs/hosts.md).
+The above example in the hosts directory pattern looks as follows.
+The `/etc/containerd/config.toml` file has the following section:
+```toml
```
Comment (rfranzke) [inline docs/usage/containerd-registry-configuration.md:37]:
```suggestion

```toml
```

File: docs/usage/containerd-registry-configuration.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+The following hosts directory structure has to be created:
+```
+$ tree /etc/containerd/certs.d
+/etc/containerd/certs.d
+└── docker.io
+    └── hosts.toml
+```
+
+Finally, for the `docker.io` upstream registry, we configure a `hosts.toml` file as follows:
+```toml
```
Comment (rfranzke) [inline docs/usage/containerd-registry-configuration.md:54]:
```suggestion

```toml
```

File: docs/usage/containerd-registry-configuration.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+server = "https://registry-1.docker.io"
+
+[host."http://public-mirror.example.com"]
+  capabilities = ["pull", "resolve"]
+```
+
+> Note: The hosts directory pattern is available in `containerd` 1.5+.
```
Comment (rfranzke) [inline docs/usage/containerd-registry-configuration.md]: You already stated this before - duplication

File: docs/usage/containerd-registry-configuration.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+### Configuring `containerd` Registries for a Shoot
+
+> Note: The below-described functionality is provided by the `ContainerdRegistryHostsDir` feature gate in gardenlet.
+
+Gardener supports configuring `containerd` registries on a Shoot using the new [hosts directory pattern](https://github.com/containerd/containerd/blob/main/docs/hosts.md). For each Shoot Node, Gardener creates the `/etc/containerd/certs.d` directory and adds the following section to the containerd's `/etc/containerd/config.toml` file:
+```toml
```
Comment (rfranzke) [inline docs/usage/containerd-registry-configuration.md:67]:
```suggestion

```toml
```

File: docs/usage/containerd-registry-configuration.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+[plugins."io.containerd.grpc.v1.cri".registry] # gardener-managed
+   config_path = "/etc/containerd/certs.d"
+```
+
+This allows Shoot owners to use the [hosts directory pattern](https://github.com/containerd/containerd/blob/main/docs/hosts.md) to configure registries for containerd. To do this, the Shoot owners need to create a directory under `/etc/containerd/certs.d` that is named with the upstream registry host name. In the newly created directory, a `hosts.toml` file needs to be created. For more details, see the [hosts directory pattern section](#hosts-directory-pattern) and the [upstream documentation](https://github.com/containerd/containerd/blob/main/docs/hosts.md).
+
+### The registry-cache Extension
+
+[Configuring `containerd` registries for a Shoot](#configuring-containerd-registries-for-a-shoot) is not the recommended approach for configuring a pull through cache for a Shoot. There is a Gardener-native extension named [registry-cache](https://github.com/gardener/gardener-extension-registry-cache) that manages a pull through cache for a Shoot using the upstream [distribution/distribution](https://github.com/distribution/distribution) project.
```
Comment (rfranzke) [inline docs/usage/containerd-registry-configuration.md]: In the light of L76, it probably is for now? 😄

File: docs/usage/containerd-registry-configuration.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+> Note: The [registry-cache](https://github.com/gardener/gardener-extension-registry-cache) extension is currently under active development and not recommended for productive usage.
+
+### Migration
+
+This section describe the migration process from the old and deprecated pattern to the hosts directory pattern for a Shoot cluster.
+
+Let's assume that the following `containerd` registries configuration using the old and deprecated pattern is being configured (for example via DaemonSet) for a Shoot:
+```toml
```
Comment (rfranzke) [inline docs/usage/containerd-registry-configuration.md:86]:
```suggestion

```toml
```

File: docs/usage/containerd-registry-configuration.md [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+version = 2
+
+[plugins."io.containerd.grpc.v1.cri".registry]
+   config_path = "/etc/containerd/certs.d"
+```
+
+1.2 The appropriate directory structure and `hosts.toml` file has to be created as described in the [hosts directory pattern section](#hosts-directory-pattern).
+
+2. When the `ContainerdRegistryHostsDir` feature gate is GA, then the machinery that performs step 1.1 can be removed. A Shoot cluster can rely that the `config_path` will be always set by gardenlet.
```
Comment (rfranzke) [inline docs/usage/containerd-registry-configuration.md]:
```suggestion
1. The `containerd` registries configuration has to be adapted to the hosts directory pattern.
1.1 The `/etc/containerd/config.toml` file needs to be adapted as follows:
    ```toml
    version = 2

    [plugins."io.containerd.grpc.v1.cri".registry]
       config_path = "/etc/containerd/certs.d"
    ```
1.2 The appropriate directory structure and `hosts.toml` file has to be created as described in the [hosts directory pattern section](#hosts-directory-pattern).
2. When the `ContainerdRegistryHostsDir` feature gate is GA, then the machinery that performs step 1.1 can be removed. A Shoot cluster can rely that the `config_path` will be always set by gardenlet.
```

File: pkg/component/extensions/operatingsystemconfig/operatingsystemconfig_test.go [hunk 15]
```diff
@@ -322,6 +322,11 @@ var _ = Describe("OperatingSystemConfig", func() {
 
 		Describe("#Deploy", func() {
 			It("should successfully deploy the shoot access secret for the cloud config downloader", func() {
+				defer test.WithVars(
+					&DownloaderConfigFn, downloaderConfigFn,
+					&OriginalConfigFn, originalConfigFn,
+				)()
```
Comment (rfranzke) [inline pkg/component/extensions/operatingsystemconfig/operatingsystemconfig_test.go:328]: Why is this necessary? I don't see other changes in this file 👀

File: pkg/provider-local/webhook/controlplane/ensurer.go [hunk 16]
```diff
@@ -163,6 +164,32 @@ Requires=` + unitNameInitializer,
...
+		// appendFileIfNotPresent in used instead of appendUniqueFile intentionally to allow enabling and testing the registry-cache extension in local setup.
+		// A file appended by the registry-cache extension is always picked up because:
+		// - if a file is already appended by the registry-cache extension, provider-local won't overwrite it (appendFileIfNotPresent)
+		// - if a file is already appended by provider-local, the registry-cache extension will overwrite it (appendUniqueFile)
+		appendFileIfNotPresent(new, extensionsv1alpha1.File{
+			Path:        filepath.Join("/etc/containerd/certs.d", mirror.UpstreamHost, "hosts.toml"),
+			Permissions: pointer.Int32(0644),
+			Content: extensionsv1alpha1.FileContent{
+				Inline: &extensionsv1alpha1.FileContentInline{
+					Encoding: "",
```
Comment (rfranzke) [inline pkg/provider-local/webhook/controlplane/ensurer.go]:
```suggestion
```

File: pkg/provider-local/webhook/controlplane/controlplane_suite_test.go [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+func TestControlPlaneWebhook(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "ControlPlane Webhook Suite")
```
Comment (rfranzke) [inline pkg/provider-local/webhook/controlplane/controlplane_suite_test.go]:
```suggestion
	RunSpecs(t, "Provider-Local Webhook ControlPlane Suite")
```

File: pkg/provider-local/webhook/controlplane/templates/scripts/configure-containerd.tpl.sh [hunk 18]
```diff
@@ -13,26 +13,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-
-hostname=garden.local.gardener.cloud
-
+# TODO(alidzhikov): The following code is migration code to ensure graceful enablement of the ContainerdRegistryHostsDir feature 
+# for existing Shoots with provider-local. Remove this script and the related units/files in v1.78.
```
Comment (rfranzke) [inline pkg/provider-local/webhook/controlplane/templates/scripts/configure-containerd.tpl.sh]: Why do we need this? We don't build migration code for local shoots/provider-local, typically.

Comment (ialidzhikov) [reply pkg/provider-local/webhook/controlplane/templates/scripts/configure-containerd.tpl.sh]: We need it to make the migration process for devs graceful.
Otherwise if you `git pull origin master` and run `make gardener-up` against existing setup with existing Shoot, then the containerd for a Shoot Node will fail to start. Because the containerd config.toml file will have the keys form the old configuration pattern and new keys will be added by the containerd-initializer script. Both of the configuration approaches are not compatible and containerd will fail to start -> kubelet not ready -> Node gets replaced.
That's why in this script we clean up the old applied configuration.

File: pkg/component/extensions/operatingsystemconfig/operatingsystemconfig_test.go (hunk 15 above)
Comment (ialidzhikov) [reply pkg/component/extensions/operatingsystemconfig/operatingsystemconfig_test.go:328]: We need it because otherwise it will call the original `OriginalConfigFn` and `DownloaderConfigFn` and this will panic with reason that feature gate `ContainerdRegistryHostsDir` is not registered.
In this test we verify that the shoot access secret gets created. Hence, I don't think we need to call the original `OriginalConfigFn` and `DownloaderConfigFn` funcs. As far as I see we don't test/verify something that comes from `OriginalConfigFn` and `DownloaderConfigFn`.

File: docs/usage/containerd-registry-configuration.md (hunk 12 above)
Comment (ialidzhikov) [reply docs/usage/containerd-registry-configuration.md]: You mean that the docs state that registry-cache is recommended approach for configuring a pull through cache but then the docs state that registry-cache is not recommended because it is under development yet?
I wanted to share that the pull through cache scenario will be covered by the registry-cache but on the other side we don't have a production ready release yet.
If you want I can adapt such as:
```suggestion
[Configuring `containerd` registries for a Shoot](#configuring-containerd-registries-for-a-shoot) won't be the recommended approach for configuring a pull through cache for a Shoot in near future. There is a Gardener-native extension named [registry-cache](https://github.com/gardener/gardener-extension-registry-cache) that manages a pull through cache for a Shoot using the upstream [distribution/distribution](https://github.com/distribution/distribution) project.
```

Comment (rfranzke) [reply docs/usage/containerd-registry-configuration.md]: Yes, this sounds more consistent at least :)

File: pkg/provider-local/webhook/controlplane/templates/scripts/configure-containerd.tpl.sh (hunk 18 above)
Comment (rfranzke) [reply pkg/provider-local/webhook/controlplane/templates/scripts/configure-containerd.tpl.sh]: Yes, just saying that we typically don't add additional code just for this. We can easily setup a new env in minutes with `make kind-down kind-up gardener-up` and be good. Anyways, it's built already now, so let's keep it.

**Comment by ialidzhikov** [comment]
I updated the feature gate's version to v1.77.0 as https://github.com/gardener/gardener/pull/8267 got merged meanwhile.
